File: String/Manacher.py
"""
Manacher algorithm: Find the longest palindromic substring in linear time complexity.

Time complexity: O(N), where N is the length of the string.
"""

class Solution:
    def manacher(self,s:str)->str:
        s_process = '#'+'#'.join(s)+'#' # add bogus to the string, so that we can handle both odd and even length
        pal_radius = [0 for _ in range(len(s_process))]
        center = 0
        max_right = 0 # the index of the right boundary of the longest palindrome
        max_len = 0 # the length of the longest palindrome found so far
        max_len_center = 0 # the index of the center of the longest palindrome found so far

        for i in range(len(s_process)):
            if i < max_right:
                mirror = 2*center-i
                pal_radius[i] = min(max_right - i, pal_radius[mirror])
            
            # attempt to expand around the current center i
            left = i - (pal_radius[i]+1)
            right = i + (pal_radius[i]+1)

            while(left >= 0 and right < len(s_process) and s_process[left]==s_process[right]):
                pal_radius[i]+=1
                left-=1
                right+=1
            
            # check if the palindrome centered at i extends beyond the right boundary
            if i+pal_radius[i] > max_right:
                center = i
                max_right = i+pal_radius[i]

            # update the longest palindrome found so far
            if pal_radius[i] > max_len:
                max_len = pal_radius[i]
                max_len_center = i

        # extract the longest palindromic substring
        start = (max_len_center - max_len) // 2
        end = start + max_len
        res = s[start:end]
        return res

if __name__=="__main__":
    str = input()
    s = Solution()
    print(s.manacher(str),end=" ") 

File: String/Manacher.cs
/*
Manacher algorithm: Find the longest palindromic substring in linear time complexity.

Time complexity: O(N), where N is the length of the string.
*/

using System;
using System.Collections.Generic;

class Solution{
    public static string manacher(string s){
        string s_prime = "";
        foreach(char x in s){
            s_prime+= x+"#";
        }
        s_prime = "#"+s_prime;
        int[] pal_radius = new int[s_prime.Length];
        Array.Fill(pal_radius,0);
        int center = 0;
        int right_bound = 0;
        int max_len = 0;
        int center_max_len = 0;
        for(int i = 0; i < s_prime.Length; ++i){
            if(i < right_bound){
                int mirror = 2*center-i;
                pal_radius[i] = Math.Min(right_bound-i,pal_radius[mirror]);
            }
            int l = i-(pal_radius[i]+1);
            int r = i+(pal_radius[i]+1);
            while(l >= 0 && r < s_prime.Length && s_prime[l]==s_prime[r]){
                pal_radius[i]++;
                l--;
                r++;
            }
            if(i+pal_radius[i] > right_bound){
                center = i;
                right_bound = i+pal_radius[i];
            }
            if(pal_radius[i] > max_len){
                max_len = pal_radius[i];
                center_max_len = i;
            }
        }
        int start = (center_max_len - max_len) >> 1;
        int end = start + max_len;
        string res = s.Substring(start,end-start);
        return res;
    }
    public static int Main(){
        string s = "abbacd";
        Console.Write(manacher(s));
        return 0;
    }
}
